node/src/components/consensus/era_supervisor.rs

```diff
@@ -180,45 +183,51 @@ where
 
         let bonded_eras = era_supervisor.bonded_eras();
         let era_ids: Vec<EraId> = era_supervisor
-            .iter_past(current_era, era_supervisor.bonded_eras().saturating_mul(3))
+            .iter_past(current_era, bonded_eras.saturating_mul(3))
             .collect();
 
-        // Asynchronously collect the information needed to initialize all recent eras.
-        let effects = async move {
-            info!(?era_ids, "collecting key blocks and booking blocks");
-
-            let key_blocks = effect_builder
-                .collect_key_block_headers(era_ids.iter().cloned())
-                .await
-                .expect("should have all the key blocks in storage");
-
-            let booking_blocks = collect_booking_block_hashes(
-                effect_builder,
-                era_ids.clone(),
-                auction_delay,
-                activation_era_id,
-            )
-            .await;
-
-            if current_era > activation_era_id.saturating_add(bonded_eras.saturating_mul(2).into())
-            {
-                // All eras can be initialized using the key blocks only.
-                (key_blocks, booking_blocks, Default::default())
-            } else {
-                let activation_era_validators = effect_builder
-                    .get_era_validators(activation_era_id)
-                    .await
-                    .unwrap_or_default();
-                (key_blocks, booking_blocks, activation_era_validators)
+        // Collect the information needed to initialize all recent eras.
+        let mut key_blocks = HashMap::new();
+        let mut booking_blocks = HashMap::new();
+        for era_id in era_ids {
+            let minimum_era = era_supervisor
+                .protocol_config
+                .last_emergency_restart
+                .unwrap_or_else(|| EraId::from(0u64));
+            let key_block_era_id = era_id.saturating_sub(1);
+            if key_block_era_id < minimum_era {
+                continue;
+            }
+            let block_header = storage
+                .read_switch_block_header_by_era_id(key_block_era_id)?
+                .ok_or_else(|| anyhow::Error::msg("No such switch block"))?;
+            key_blocks.insert(era_id, block_header);
+            let booking_block_era_id = era_id
+                .saturating_sub(era_supervisor.protocol_config.auction_delay)
+                .saturating_sub(1);
+            if booking_block_era_id < minimum_era {
+                continue;
             }
```

Comment:
Have you tried running an emergency restart test? Because I'm pretty sure this would cause it to fail.

Consider this scenario: we are starting right after an emergency restart, the last era before was `N-1`, now we created a dummy switch block for era `N` and are trying to initialize era `N+1`. `minimum_era` will be `N`. `booking_block_era_id` will be `N - auction_delay`, so it's less than `minimum_era`, so we'll hit the `continue` statement and won't save a booking block for era `N+1`. But then in `handle_initialize_eras` we panic if there's no booking block for an era - which means we'll panic.

Am I missing something?

Reply:
Fixed in c2fa0b4b5